File: jarvis/brain.py
"""The assistant's brain: a Claude conversation loop with tool use."""
from typing import Callable, Optional

from anthropic import Anthropic

from jarvis.config import SETTINGS
from jarvis.tools import TOOLS, run_tool

SYSTEM_PROMPT_TEMPLATE = """You are Jarvis, a witty, imperturbably composed personal AI assistant \
in the mold of Tony Stark's J.A.R.V.I.S. You address the user as "{user_name}". \
Keep spoken responses concise (a sentence or two) since they may be read aloud by \
text-to-speech — save detail for when the user asks for it. Use your tools whenever \
a question needs current information, or the user asks you to manage tasks, \
reminders, notes, or their system, rather than guessing. Never invent tool results."""


class Jarvis:
    def __init__(
        self,
        tools: Optional[list[dict]] = None,
        tool_runner: Optional[Callable[[str, dict], str]] = None,
        system_prompt: Optional[str] = None,
    ):
        SETTINGS.require_api_key()
        self.client = Anthropic(api_key=SETTINGS.anthropic_api_key)
        self.model = SETTINGS.model
        self.system_prompt = system_prompt or SYSTEM_PROMPT_TEMPLATE.format(user_name=SETTINGS.user_name)
        self.tools = tools if tools is not None else TOOLS
        self.tool_runner = tool_runner
        self.history: list[dict] = []

    def reset(self):
        self.history = []

    def ask(self, user_text: str, context: str = "") -> str:
        """Send one user turn and run the tool loop until Claude answers in text.

        `context` is appended to the system prompt for this call only (e.g. the
        phone's local time), so it never piles up in the conversation history.
        """
        runner = self.tool_runner or run_tool
        system = f"{self.system_prompt}\n\n{context}" if context else self.system_prompt
        self.history.append({"role": "user", "content": user_text})

        while True:
            response = self.client.messages.create(
                model=self.model,
                max_tokens=1024,
                system=system,
                tools=self.tools,
                messages=self.history,
            )
            self.history.append({"role": "assistant", "content": response.content})

            if response.stop_reason != "tool_use":
                return "".join(
                    block.text for block in response.content if block.type == "text"
                ).strip()

            tool_results = []
            for block in response.content:
                if block.type == "tool_use":
                    result = runner(block.name, block.input)
                    tool_results.append(
                        {
                            "type": "tool_result",
                            "tool_use_id": block.id,
                            "content": result,
                        }
                    )
            self.history.append({"role": "user", "content": tool_results})

File: android/core/src/main/kotlin/com/jarvis/core/Brain.kt
package com.jarvis.core

import com.anthropic.client.AnthropicClient
import com.anthropic.core.JsonValue
import com.anthropic.models.messages.CacheControlEphemeral
import com.anthropic.models.messages.ContentBlockParam
import com.anthropic.models.messages.Message
import com.anthropic.models.messages.MessageCreateParams
import com.anthropic.models.messages.MessageParam
import com.anthropic.models.messages.OutputConfig
import com.anthropic.models.messages.StopReason
import com.anthropic.models.messages.TextBlockParam
import com.anthropic.models.messages.ToolResultBlockParam
import com.anthropic.models.messages.WebSearchTool20260209
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale

const val DEFAULT_MODEL = "claude-opus-5"

/**
 * The Claude conversation loop: sends the user's words plus Jarvis's tools,
 * runs any tools Claude asks for, and repeats until Claude answers in text.
 */
class Brain(
    private val client: AnthropicClient,
    private val tools: JarvisTools,
    private val model: String = DEFAULT_MODEL,
    private val userName: String = "sir",
    private val clock: () -> ZonedDateTime = { ZonedDateTime.now() },
) {
    private val systemPrompt = """
        You are Jarvis, a witty, imperturbably composed personal AI assistant in the mold of Tony Stark's J.A.R.V.I.S., running on the user's Android phone. Address the user as "$userName".
        Your replies are spoken aloud: keep them to one or two short sentences unless asked for detail, and never use markdown, lists, links or emoji. Latency-sensitive; begin your visible answer immediately.
        Use your tools rather than guessing: web_search for anything current (news, weather, scores, prices), the phone tools to open apps and websites, call, text, WhatsApp, navigate, play things on YouTube, set alarms and timers, and control the flashlight and volume, and the task, note and reminder tools to keep track of things.
        To call or text someone by name, look them up with find_contact first; if there are several numbers, ask which one.
        Each user message starts with the phone's local time in brackets; use it for reminders and anything time-related.
        Never claim you did something unless a tool result says it happened.
    """.trimIndent()

    private val history = mutableListOf<MessageParam>()
    private var lastAsk: ZonedDateTime? = null

    @Synchronized
    fun reset() {
        history.clear()
    }

    /** Sends one user turn and returns Jarvis's spoken reply. Throws on network/API errors. */
    @Synchronized
    fun ask(userText: String): String {
        val now = clock()
        // Start fresh after a long pause or a long conversation: keeps requests
        // cheap and fast, and old context rarely matters to a voice assistant.
        if (lastAsk?.plusMinutes(15)?.isBefore(now) == true || history.size > 40) history.clear()
        lastAsk = now

        val startSize = history.size
        history.add(userMessage("[${now.format(TIME_FORMAT)}]\n$userText"))
        try {
            repeat(MAX_ROUNDS) {
                val response = client.messages().create(buildParams())
                val stop = response.stopReason().orElse(null)

                if (stop == StopReason.REFUSAL) {
                    rollback(startSize)
                    return "I'm afraid I can't help with that one, $userName."
                }
                if (stop == StopReason.MAX_TOKENS) {
                    rollback(startSize)
                    return "That answer ran too long. Ask me for a shorter version."
                }

                assistantMessage(response)?.let { history.add(it) }

                when (stop) {
                    // Server-side web search paused a long turn; send it back to continue.
                    StopReason.PAUSE_TURN -> return@repeat
                    StopReason.TOOL_USE -> history.add(runTools(response))
                    else -> return spokenText(response)
                }
            }
            rollback(startSize)
            return "That took too many steps. Could you try asking another way?"
        } catch (e: Exception) {
            // Leave history exactly as it was so the next request is still valid.
            rollback(startSize)
            throw e
        }
    }

    private fun buildParams(): MessageCreateParams {
        val builder = MessageCreateParams.builder()
            .model(model)
            .maxTokens(4096L)
            .systemOfTextBlockParams(
                listOf(TextBlockParam.builder().text(systemPrompt).cacheControl(CacheControlEphemeral.builder().build()).build()),
            )
            .messages(history.toList())
            // Voice assistant: fast answers matter more than deep deliberation.
            .outputConfig(OutputConfig.builder().effort(OutputConfig.Effort.LOW).build())
            .addTool(WebSearchTool20260209.builder().maxUses(3L).build())
            // If Claude Opus 5 declines a request, the API retries it on
            // Anthropic's recommended fallback model instead of refusing.
            .putAdditionalHeader("anthropic-beta", "server-side-fallback-2026-07-01")
            .putAdditionalBodyProperty("fallbacks", JsonValue.from("default"))
        tools.definitions.forEach { builder.addTool(it) }
        return builder.build()
    }

    private fun runTools(response: Message): MessageParam {
        val results = response.content().mapNotNull { block -> block.toolUse().orElse(null) }.map { call ->
            val input = runCatching { call._input().convert(Map::class.java) }.getOrNull() ?: emptyMap<String, Any>()
            val (result, isError) = tools.run(call.name(), input)
            ContentBlockParam.ofToolResult(
                ToolResultBlockParam.builder().toolUseId(call.id()).content(result).isError(isError).build(),
            )
        }
        return MessageParam.builder().role(MessageParam.Role.USER).contentOfBlockParams(results).build()
    }

    private fun rollback(size: Int) {
        while (history.size > size) history.removeAt(history.size - 1)
    }

    private companion object {
        const val MAX_ROUNDS = 8
        val TIME_FORMAT: DateTimeFormatter = DateTimeFormatter.ofPattern("EEEE d MMMM yyyy, HH:mm, VV", Locale.ENGLISH)

        fun userMessage(text: String): MessageParam =
            MessageParam.builder().role(MessageParam.Role.USER).content(text).build()

        /** The assistant turn to replay next request, including thinking and server-tool blocks. */
        fun assistantMessage(response: Message): MessageParam? {
            // Blocks this SDK version can't convert (e.g. newer block types) are skipped.
            val blocks = response.content().mapNotNull { runCatching { it.toParam() }.getOrNull() }
            if (blocks.isEmpty()) return null
            return MessageParam.builder().role(MessageParam.Role.ASSISTANT).contentOfBlockParams(blocks).build()
        }

        fun spokenText(response: Message): String =
            response.content().mapNotNull { it.text().orElse(null)?.text() }.joinToString(" ").trim()
                .ifEmpty { "Done." }
    }
}
